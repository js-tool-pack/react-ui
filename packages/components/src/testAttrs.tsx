import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
