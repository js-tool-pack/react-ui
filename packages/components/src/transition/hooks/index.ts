export * from './useDispatcher';
export * from './useDispatcher';
export * from './useTransition';
