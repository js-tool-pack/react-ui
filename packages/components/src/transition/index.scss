@use '../namespace' as Name;

$r: Name.$transition;

.#{$r} {
  &--invisible {
    display: none !important;
  }
}
