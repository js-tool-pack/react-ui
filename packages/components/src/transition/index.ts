export type { TransitionProps, CB as TransitionCB } from './transition.types';
export { default as Transition } from './Transition';
export {
  STATUS as TRANSITION_STATUS,
  LIFE_CIRCLE as TRANSITION_LIFE_CIRCLE,
} from './transition.enums';
export { transitionCBAdapter } from './transition.utils';
