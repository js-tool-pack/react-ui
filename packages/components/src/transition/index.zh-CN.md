---
order: -2
category: Components
title: Transition 动画
atomId: Transition
cover: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*BrFMQ5s7AAQAAAAAAAAAAAAADrJ8AQ/original
coverDark: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*Lp1kTYmSsgoAAAAAAAAAAAAADrJ8AQ/original
demo:
  cols: 2
group:
  order: 1
  title: 动画
---

Transition 会在一个元素或组件进入和离开 DOM 时应用动画，类似于 Vue 官方的 Transition 组件。

## 何时使用

页面元素需要添加入场、退场动画时。

## className 生命周期

当设 name 为 fade 时

### 入场

`.fade-enter-from` -> `.fade-enter-active` -> `.fade-enter-active.fade-enter-to` -> 动画完成后清理相关 className

### 退场

`.fade-leave-from` -> `.fade-leave-active` -> `.fade-leave-active.fade-leave-to` -> 动画完成后清理相关 className

## children

单个 children 与多个 children 的逻辑有点不同：
:::info{title=children}
多个 children 相对于单个 children 需要为每个 children 设置 key 属性以便 Transition 组件识别组件是否变动，  
且可以通过 mode 设置入场、退场的顺序
:::

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/all.tsx"></code>
<code src="./demo/basic.tsx"></code>
<code src="./demo/appear.tsx"></code>
<code src="./demo/mode.tsx"></code>
<code src="./demo/on.tsx"></code>
<code src="./demo/adpter.tsx"></code>
<code src="./demo/input.tsx"></code>
<code src="./demo/show.tsx"></code>
<code src="./demo/show-appear.tsx"></code>

## API

Transition 的属性说明如下：

| 属性   | 说明                                                   | 类型                              | 默认值    | 版本 |
| ------ | ------------------------------------------------------ | --------------------------------- | --------- | ---- |
| name   | 动画名，用于生成 className                             | string                            | `trans`   |      |
| mode   | 动画类型，用于决定子元素进出场顺序                     | `out-in` \| `in-out` \| `default` | `default` |      |
| appear | 用于设置初始时是否显示动画，只对组件第一次出现时起作用 | boolean                           | false     |
| on     | 切换动画时的回调                                       | TransitionCB                      | -         |      |
| show   | 类似 vue 版的 v-show 控制动画                          | boolean \| undefined              | -         |      |
