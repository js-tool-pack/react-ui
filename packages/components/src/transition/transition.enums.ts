export enum LIFE_CIRCLE {
  ready,
  before,
  run,
  running,
  start,
  after,
  cancel,
}

export enum STATUS {
  none,
  idle,
  show,
  hide,
  invisible,
}
