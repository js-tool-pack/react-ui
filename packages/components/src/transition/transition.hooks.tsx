import { cloneElement, useCallback, useEffect, useMemo, useRef } from 'react';
import { addTransition, getClasses } from './transition.utils';
import { LIFE_CIRCLE, STATUS } from './transition.enums';
import type { CB, El, Mode } from './transition.types';
import { useForceUpdate, useIsInitDep } from '@pkg/shared';

export function useDispatcher(
  mode: Mode,
  // show?: boolean,
  appear: boolean,
  children: El,
  cb?: CB,
) {
  const forceUpdate = useForceUpdate();
  const cbRef = useRef<CB | void>();
  const isInitDep = useIsInitDep(children);

  const [prev, next] = useChildren(children);
  const prevRef = useRef<El>();
  const nextRef = useRef<El>();
  const prevStatusRef = useRef<STATUS>(STATUS.none);
  const nextStatusRef = useRef<STATUS>(STATUS.none);

  const setChild = useCallback((p?: El, n?: El) => {
    prevRef.current = p;
    nextRef.current = n;
    // forceUpdate();
  }, []);
  const setStatus = useCallback((p: STATUS, n: STATUS) => {
    prevStatusRef.current = p;
    nextStatusRef.current = n;
    forceUpdate();
  }, []);
  const setVisible = useCallback((el: HTMLElement) => {
    const originVis = el.style.visibility;
    el.style.visibility = 'hidden';
    setTimeout(() => (el.style.visibility = originVis));
  }, []);
  const createAfterHandler = useCallback(
    (p = STATUS.none, n = STATUS.idle) => {
      return ((el, status, lifeCircle) => {
        cb?.(el, status, lifeCircle);
        if (status === STATUS.hide && LIFE_CIRCLE.after === lifeCircle) {
          setVisible(el);
          setStatus(p, n);
          cbRef.current = undefined;
        }
      }) as CB;
    },
    [cb, setStatus, setVisible],
  );

  useEffect(() => {
    setChild(prev, next);

    if (isInitDep && !appear) {
      setStatus(STATUS.idle, STATUS.idle);
      return;
    }

    if ((prev && !next) || (!prev && next)) {
      setStatus(STATUS.hide, STATUS.show);
      cbRef.current = createAfterHandler();
    } else {
      switch (mode) {
        case 'out-in':
          setStatus(STATUS.hide, STATUS.none);
          cbRef.current = createAfterHandler(STATUS.none, STATUS.show);
          break;
        case 'in-out':
          setStatus(STATUS.idle, STATUS.show);
          cbRef.current = (_, __, lifeCircle) => {
            if (LIFE_CIRCLE.after === lifeCircle) {
              setStatus(STATUS.hide, STATUS.idle);
              cbRef.current = createAfterHandler();
            }
          };
          break;
        default:
          setStatus(STATUS.hide, STATUS.show);
          cbRef.current = createAfterHandler();
          break;
      }
    }
    return () => {
      setStatus(STATUS.none, STATUS.none);
      cbRef.current = undefined;
    };
  }, [prev, next]);

  return [
    prevRef.current,
    nextRef.current,
    prevStatusRef.current,
    nextStatusRef.current,
    cbRef.current,
  ] as const;
}

export function useChildren<T extends El[] | El>(children: T) {
  const prev = useRef<T>();
  const memo = useMemo<[T | void, T | void]>(
    () => [prev.current, children],
    [children],
  );
  useEffect(() => {
    prev.current = children;
  }, [children]);
  return memo;
}

export function useTransition(
  // from: string,
  status: STATUS,
  name: string,
  children?: El,
  cb?: void | CB,
) {
  const elRef = useRef<HTMLElement>(null);

  useEffect(() => {
    const el = elRef.current;
    // console.log(from, 'status', STATUS[status], !!transRef.current);
    if (!children || !el || status === STATUS.none) return;
    // console.log(from, 'status', STATUS[status], !!transRef.current);

    const classes = getClasses(name, status === STATUS.show);

    const trans = addTransition({
      el,
      classes,
      on: (lifeCircle) => {
        // console.log('[life circle]:', LIFE_CIRCLE[type]);
        cb?.(el, status, lifeCircle);
      },
    });

    el.style.visibility = '';
    trans.start();
    return () => {
      trans.clearListener();
      // 如果把active的class也清理掉就跟vue的差不多了，不过那种动画会从头开始
      trans.removeClass();
    };
  }, [children, status, name, elRef]);

  if (!children || STATUS.none === status) return;
  if (status === STATUS.idle) return children;
  return <>{cloneElement(children, { ref: elRef })}</>;
}
