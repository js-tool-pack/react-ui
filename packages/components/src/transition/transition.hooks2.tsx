import { cloneElement, useCallback, useEffect, useMemo, useRef } from 'react';
import { addTransition, getClasses } from './transition.utils';
import { LIFE_CIRCLE, STATUS } from './transition.enums';
import type { CB, El, Mode } from './transition.types';
import { nextTick, useForceUpdate, useIsInitDep } from '@pkg/shared';

export function useDispatcher(
  mode: Mode,
  // show?: boolean,
  appear: boolean,
  children: El,
  cb?: CB,
) {
  console.log('useDispatcher');
  const forceUpdate = useForceUpdate();
  const isInitDep = useIsInitDep(children);
  const [prev, next] = useChildren(children);

  const setVisible = useCallback((el: HTMLElement) => {
    const originVis = el.style.visibility;
    el.style.visibility = 'hidden';
    setTimeout(() => (el.style.visibility = originVis));
  }, []);

  const createAfterHandler = (p = STATUS.none, n = STATUS.idle) => {
    console.log('00000', STATUS[p], STATUS[n]);
    return ((el, status, lifeCircle) => {
      cb?.(el, status, lifeCircle);
      // console.log('zzzzzzz', LIFE_CIRCLE[lifeCircle]);
      if (status === STATUS.hide && LIFE_CIRCLE.after === lifeCircle) {
        setVisible(el);
      }
      if (LIFE_CIRCLE.after === lifeCircle) {
        cache.current = [p, n];
        cbRef.current = undefined;
        forceUpdate();
      }
    }) as CB;
  };

  const cache = useRef<[STATUS, STATUS]>();
  const cbRef = useRef<CB>();

  const res = (): [STATUS, STATUS] => {
    const c = cache.current;
    console.log(22222, c, prev, next);
    if (c) {
      const cc = [...c] as [STATUS, STATUS];
      // 直接设置为undefined会刷新dom失败，hide后也会显示dom
      // 需要异步才行
      // cache.current = undefined;
      nextTick(() => (cache.current = undefined));
      return cc;
    }
    if (isInitDep && !appear) {
      return [STATUS.idle, STATUS.idle];
    }
    if ((prev && !next) || (!prev && next)) {
      cbRef.current = createAfterHandler();
      return [STATUS.hide, STATUS.show];
    }
    switch (mode) {
      case 'out-in':
        cbRef.current = createAfterHandler(STATUS.none, STATUS.show);
        return [STATUS.hide, STATUS.none];
      case 'in-out':
        cbRef.current = (_, __, lifeCircle) => {
          if (LIFE_CIRCLE.after === lifeCircle) {
            cbRef.current = createAfterHandler();
            cache.current = [STATUS.hide, STATUS.idle];
            forceUpdate();
          }
        };
        return [STATUS.idle, STATUS.show];
      default:
        cbRef.current = createAfterHandler();
        return [STATUS.hide, STATUS.show];
    }
  };

  return [prev, next, ...res(), cbRef.current] as const;
}

export function useChildren<T extends El[] | El>(children: T) {
  const prev = useRef<T>();
  const memo = useMemo<[T | void, T | void]>(
    () => [prev.current, children],
    [children],
  );
  useEffect(() => {
    prev.current = children;
  }, [children]);
  return memo;
}

export function useTransition(
  // from: string,
  status: STATUS,
  name: string,
  children?: El,
  cb?: void | CB,
) {
  const elRef = useRef<HTMLElement>(null);

  const classes = getClasses(name, status === STATUS.show);

  useEffect(() => {
    const el = elRef.current;
    // console.log('=====', STATUS[status], el);
    // console.log(from, 'status', STATUS[status], !!transRef.current);
    if (!children || !el || status === STATUS.none) return;
    const trans = addTransition({
      el,
      classes,
      on: (lifeCircle) => {
        // console.log('[life circle]:', LIFE_CIRCLE[type]);
        cb?.(el, status, lifeCircle);
      },
    });
    trans.start();
    return () => {
      trans.clearListener();
      // 如果把active的class也清理掉就跟vue的差不多了，不过那种动画会从头开始
      // trans.removeClass();
    };
  }, [children]);

  console.log('ccccc', STATUS[status], children);

  if (!children || STATUS.none === status || typeof children === 'boolean')
    return;
  if (status === STATUS.idle) return children;

  const className =
    status === STATUS.show
      ? [children.props.className, classes.fromClassName]
          .filter((i) => Boolean(i))
          .join(' ')
      : children.props.className;

  return cloneElement(children, {
    ref: elRef,
    className,
  });
}
