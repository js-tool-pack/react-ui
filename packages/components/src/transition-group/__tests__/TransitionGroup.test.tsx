import { render } from '@testing-library/react';
import { Transition } from '~/transition';
import { testAttrs } from '~/testAttrs';
import { TransitionGroup } from '..';

describe('TransitionGroup', () => {
  testAttrs(TransitionGroup);

  test('snap', () => {
    const { container } = render(
      <TransitionGroup>
        <div>foo bar</div>
        <div>foo-bar</div>
      </TransitionGroup>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test('transition snap', () => {
    const { container } = render(
      <TransitionGroup>
        <Transition appear>
          <div>foo bar</div>
        </Transition>
        <Transition>
          <div>foo-bar</div>
        </Transition>
      </TransitionGroup>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
