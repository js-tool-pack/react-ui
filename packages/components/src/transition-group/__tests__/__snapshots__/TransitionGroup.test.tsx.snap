// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionGroup snap 1`] = `
<div
  class="t-transition-group"
>
  <div
    class=""
  >
    foo bar
  </div>
  <div
    class=""
  >
    foo-bar
  </div>
</div>
`;

exports[`TransitionGroup transition snap 1`] = `
<div
  class="t-transition-group"
>
  <div
    class="t-group-enter-active t-group-enter-to"
  >
    foo bar
  </div>
  <div
    class=""
  >
    foo-bar
  </div>
</div>
`;
